### ISSUE TITLE
Inconsistency between Tensorflow and Theano/CNTK when using NASNetLarge/NASNetMobile and MobileNetV2 in Keras 2.2.2

### ISSUE THREAD

**Issue by hvpham** [keras-team/keras-applications#27]
I am using the NASNetLarge/NASNetMobile and the MobileNetV2 on Keras 2.2.2. I am testing with all three backends and I see the differences between TensorFlow and Theano/CNTK.

I reported that [here](https://github.com/keras-team/keras/issues/10672)

At first, I thought it was because of the Depthwise and Separately Conv2D that was implemented for CNTK and Theano in their backend files. But later I found out that it seems to #be because of padding SAME inconsistency between TensorFlow and the other two backends.

I have seen fixes in the Mobile and ResNet50 that simply add ZeroPadding and use padding VALID. [#9473](https://github.com/keras-team/keras/pull/9473)?

Can someone confirm that we can do the same for these networks?  Or should we actually implement the padding scheme SAME of TensorFlow on Theano and CNTK to be consistent?

**Comment by taehoonlee** [comment]
Thank you for the reports, @hvpham. The results with the [`cat.png`](https://homepages.cae.wisc.edu/~ece533/images/) 🐈 are:

- `MobileNetV2`:
  - TF: `[(u'n02124075', u'Egyptian_cat', 0.23030917), (u'n02123045', u'tabby', 0.13059248)]`
  - TH: `[(u'n02123045', u'tabby', 0.15593995), (u'n03482405', u'hamper', 0.082497254)]`
  - C: `[(u'n02123045', u'tabby', 0.15593976), (u'n03482405', u'hamper', 0.08249792)]`
- `NASNetMobile`:
  - TF: `[(u'n02124075', u'Egyptian_cat', 0.3190641), (u'n02123159', u'tiger_cat', 0.24479342)]`
  - TH: `[(u'n03482405', u'hamper', 0.25978553), (u'n02124075', u'Egyptian_cat', 0.16284499)]`
  - C: `[(u'n03482405', u'hamper', 0.2597867), (u'n02124075', u'Egyptian_cat', 0.16284464)]`
- `NASNetLarge`:
  - TF: `[(u'n03482405', u'hamper', 0.7357966), (u'n02123045', u'tabby', 0.06878438)]`
  - TH: `[(u'n03482405', u'hamper', 0.9320255), (u'n04204238', u'shopping_basket', 0.014910409)]`

As you expected, the way like keras-team/keras#9473 helps the problem. Now the results with the [recent fix](https://github.com/keras-team/keras-applications/commit/b07b0b17a4056e475ac005a563ca56c5396149cd) are consistent as follows:

- `MobileNetV2`:
  - TF: `[(u'n02124075', u'Egyptian_cat', 0.23030917), (u'n02123045', u'tabby', 0.13059248)]`
  - TH: `[(u'n02124075', u'Egyptian_cat', 0.23030858), (u'n02123045', u'tabby', 0.13059291)]`
  - C: `[(u'n02124075', u'Egyptian_cat', 0.23030911), (u'n02123045', u'tabby', 0.13059254)]`
- `NASNetMobile`:
  - TF: `[(u'n02124075', u'Egyptian_cat', 0.32422182), (u'n02123159', u'tiger_cat', 0.25173983)]`
  - TH: `[(u'n02124075', u'Egyptian_cat', 0.324222), (u'n02123159', u'tiger_cat', 0.25173998)]`
  - C: `[(u'n02124075', u'Egyptian_cat', 0.3242214), (u'n02123159', u'tiger_cat', 0.25174025)]`
- `NASNetLarge`:
  - TF: `[(u'n03482405', u'hamper', 0.8357486), (u'n02123045', u'tabby', 0.037017588)]`
  - TH: `[(u'n03482405', u'hamper', 0.8357489), (u'n02123045', u'tabby', 0.037017457)]`

**Comment by taehoonlee** [comment]
@hvpham, As the inconsistency has been fixed, I'll close the issue for now. Please feel free to open it again at any time if you have further questions. I'm always ready :)

**Comment by kumardesappan** [comment]
Did you fine-tune the mobilenet weights after this fix. Since, The original TF weights will not produce same result with this fix?

**Comment by kumardesappan** [comment]
@taehoonlee Thanks for the response.
As per my understanding,
Conv2D(strides=2, padding='same') behavior is not functionally same as 
ZeroPad(Kernels/2) + Conv2D(strides=2, padding='valid') in TF. (But, it is same in other frameworks)

If I am not wrong, the TF models are trained with Conv2D(strides=2, padding='same').
You can refer the frozen models from below path. These models has Conv2D(strides=2, padding='same'). Did you use these models for translation or some other models.?

https://github.com/tensorflow/models/tree/master/research/slim/nets/mobilenet

**Comment by kumardesappan** [comment]
Actually, the fix is below if the input width is even.
ZeroPad(0, 1,0,1 ) + Conv2D(strides=2, padding='valid') 
Which is fucntioanlly equllent to Conv2D(strides=2, padding='same').
I did not notice the "correct_pad" function earlier.

**Comment by taehoonlee** [comment]
@kumardesappan, Sorry but it is hard for me to see your point... Would you please elaborate the point?
